### PR TITLE
TotalVI model addition

### LINKED ISSUE #456: Dataset10X sparse default and protein expression
There are two issues with Dataset10X:

1)

https://github.com/YosefLab/scVI/blob/ab1ac5938d1c0df84d9c6e4764b5f226044c7154/scvi/dataset/dataset10X.py#L204

This is my fault - we should first check if the dataset is already dense before densifying the protein expression

2)  We should set a threshold for automatically sparsifying/densifying a dataset like we have in other classes

### PR DESCRIPTION
**Pull request by adamgayoso**
As well as adding the model I changed a couple of things globally

- The NB log likelihood now does not return a sum, which is consistent with other pytorch distributions
- Fixes #456 an issue with 10x datasets and protein data sparsity 
- Modify `FCLayers` so that it can be used as a linear decoder, which will allow use to wrap LDVAE into VAE in the future

There was a bit of a struggle complying with the current levels of abstraction given how different this model is, that said, I'm sure in the near future we can wrap totalVI and scVI into one pytorch class

### CODE REVIEW

File: scvi/inference/total_inference.py [hunk 1]
```diff
@@ -0,0 +1,608 @@
...
+            {
+                "collate_fn": gene_dataset.collate_fn_builder(
+                    {"protein_expression": np.float32}
+                )
+            }
+        )
+        self.data_loader = DataLoader(gene_dataset, **self.data_loader_kwargs)
+
+    def corrupted(self):
+        return self.update(
```
[the remark this answers is not in the document]
Comment (adamgayoso) [reply scvi/inference/total_inference.py]: I do plan on using this.

File: scvi/inference/total_inference.py [hunk 2]
```diff
@@ -0,0 +1,608 @@
...
+
+        super().__init__(
+            model,
+            gene_dataset,
+            shuffle=shuffle,
+            indices=indices,
+            use_cuda=use_cuda,
+            data_loader_kwargs=data_loader_kwargs,
+        )
+        # Add protein tensor as another tensor to be loaded
```
[the remark this answers is not in the document]
Comment (adamgayoso) [reply scvi/inference/total_inference.py]: if `dataset.protein_expression` does not exist, this will fail. Should that be more clear?

File: scvi/inference/total_inference.py [hunk 3]
```diff
@@ -0,0 +1,608 @@
...
+            insights on the modeling of the data, and is fast to compute
+
+            This is really a helper function to self.ll, self.ll_protein, etc.
+        """
+        # Iterate once over the posterior and computes the total log_likelihood
+        log_lkl_gene = 0
+        log_lkl_protein = 0
+        for i_batch, tensors in enumerate(self):
+            x, local_l_mean, local_l_var, batch_index, labels, y = tensors
+            reconst_loss_gene, reconst_loss_protein, kl_divergence, back_kl = vae(
```
[the remark this answers is not in the document]
Comment (adamgayoso) [reply scvi/inference/total_inference.py]: I'm not sure I understand - this is how we have it for scVI in `log_likelihood.py`. I don't think there's a way to amortize computations.

File: scvi/dataset/synthetic.py [hunk 4]
```diff
@@ -36,6 +40,14 @@ def __init__(
         # clear potentially unused cell_types
         self.remap_categorical_attributes()
 
+        protein_data = CellMeasurement(
```
[the remark this answers is not in the document]
Comment (adamgayoso) [reply scvi/dataset/synthetic.py]: I updated this so that protein data is independent of gene data. I think it should be fine generating protein data by default right? This doesn't effect at all and scVI functions - it's just another dataset attribute.

File: scvi/inference/total_inference.py [hunk 5]
```diff
@@ -0,0 +1,608 @@
...
+         to use Default: ``0.02``. Note that if train and test do not add to 1 the remainder is placed in a validation set
+        :\*\*kwargs: Other keywords arguments from the general Trainer class.
+    """
+    default_metrics_to_monitor = ["elbo"]
+
+    def __init__(
+        self,
+        model,
+        dataset,
+        train_size=0.93,
```
[the remark this answers is not in the document]
Comment (adamgayoso) [reply scvi/inference/total_inference.py]: I update to 90/5/5 - issue is that datasets are small and need a lot of training data to get protein background disentangled.

File: scvi/models/modules.py [hunk 6]
```diff
@@ -475,3 +480,245 @@ def forward(
...
+            px_rate,
+            px_dropout,
+            py_back_alpha,
+            py_back_beta,
+            log_back_mean,
+        )
+
+
+# Encoder
+class EncoderTOTALVI(nn.Module):
```
[the remark this answers is not in the document]
Comment (adamgayoso) [reply scvi/models/modules.py]: this is the "more efficient" scVI encoder with parameter sharing versus the default. We can name it as such?

File: scvi/models/totalvi.py [hunk 7]
```diff
@@ -0,0 +1,541 @@
...
+
+    :param log_variational: Log(data+1) prior to encoding for numerical stability. Not normalization.
+    :param reconstruction_loss_genes:  One of
+
+        * ``'nb'`` - Negative binomial distribution
+        * ``'zinb'`` - Zero-inflated negative binomial distribution
+
+    :param reconstruction_loss_protein:  One of
+
+        * ``'mixture_nb'`` - Negative binomial mixture distribution
```
[the remark this answers is not in the document]
Comment (adamgayoso) [reply scvi/models/totalvi.py]: done.

File: scvi/models/totalvi.py [hunk 8]
```diff
@@ -0,0 +1,541 @@
...
+
+    :param latent_distribution:  One of
+
+        * ``'normal'`` - Isotropic normal
+        * ``'gsm'`` - Gaussian softmax N(0, 1) passed through softmax
```
[the remark this answers is not in the document]
Comment (adamgayoso) [reply scvi/models/totalvi.py]: done

File: scvi/models/totalvi.py [hunk 9]
```diff
@@ -0,0 +1,541 @@
...
+            )
+            self.background_log_beta = torch.nn.Parameter(torch.randn(n_input_proteins))
+
+        if self.gene_dispersion == "gene":
+            self.px_r_gene = torch.nn.Parameter(torch.randn(n_input_genes))
+        elif self.gene_dispersion == "gene-batch":
+            self.px_r_gene = torch.nn.Parameter(torch.randn(n_input_genes, n_batch))
+        elif self.gene_dispersion == "gene-label":
+            self.px_r_gene = torch.nn.Parameter(torch.randn(n_input_genes, n_labels))
+        else:  # gene-cell
```
[the remark this answers is not in the document]
Comment (adamgayoso) [reply scvi/models/totalvi.py]: I don't use it. Should I reimplement it for genes? No other reason.

File: scvi/models/totalvi.py [hunk 10]
```diff
@@ -0,0 +1,541 @@
...
+            n_latent,
+            n_input_genes,
+            self.n_input_proteins,
+            n_layers=1,
+            n_cat_list=[n_batch, n_labels],
+            n_hidden=n_hidden,
+            dropout_rate=dropout_rate_decoder,
+        )
+
+    def get_latents(self, x: torch.Tensor, batch_index: Optional[torch.Tensor] = None):
```
[the remark this answers is not in the document]
Comment (adamgayoso) [reply scvi/models/totalvi.py]: it's actually get latents in the vae class and get latent in posterior.

File: scvi/models/totalvi.py [hunk 11]
```diff
@@ -0,0 +1,541 @@
...
+        x: torch.Tensor,
+        y: torch.Tensor,
+        batch_index: Optional[torch.Tensor] = None,
+        give_mean: bool = True,
+    ) -> torch.Tensor:
+        r""" samples the tensor of library size from the posterior
+        #doesn't really sample, returns the tensor of the means of the posterior distribution
+
+        :param x: tensor of values with shape ``(batch_size, n_input_genes)``
+        :param y: tensor of cell-types labels with shape ``(batch_size, n_labels)``
```
[the remark this answers is not in the document]
Comment (adamgayoso) [reply scvi/models/totalvi.py]: done.

File: tests/test_scvi.py [hunk 12]
```diff
@@ -221,3 +242,10 @@ def test_sampling_zl(save_path):
     )
     trainer_cortex_cls.train(n_epochs=2)
     trainer_cortex_cls.test_set.accuracy()
+
+
+def test_totalvi(save_path):
+    synthetic_datset_one_batch = SyntheticDataset(n_batches=1)
+    totalvi_benchmark(synthetic_datset_one_batch, n_epochs=1, use_cuda=False)
```
[the remark this answers is not in the document]
Comment (adamgayoso) [reply tests/test_scvi.py]: done.

File: tests/test_scvi.py [hunk 13]
```diff
@@ -221,3 +242,10 @@ def test_sampling_zl(save_path):
     )
     trainer_cortex_cls.train(n_epochs=2)
     trainer_cortex_cls.test_set.accuracy()
+
+
+def test_totalvi(save_path):
+    synthetic_datset_one_batch = SyntheticDataset(n_batches=1)
```
[the remark this answers is not in the document]
Comment (adamgayoso) [reply tests/test_scvi.py]: done.

File: scvi/models/totalvi.py [hunk 14]
```diff
@@ -0,0 +1,541 @@
...
+        if self.reconstruction_loss_protein == "poisson":
+            reconst_loss_protein = (
+                -Poisson(px_rate["protein"]).log_prob(protein).sum(dim=-1)
+            )
+        if self.reconstruction_loss_protein == "nb":
+            reconst_loss_protein = -log_nb_positive(
+                protein, px_rate["protein"], px_r["protein"]
+            ).sum(dim=-1)
+        if self.reconstruction_loss_protein == "mixture_nb":
+
```
[the remark this answers is not in the document]
Comment (adamgayoso) [reply scvi/models/totalvi.py]: done.

File: scvi/models/totalvi.py [hunk 15]
```diff
@@ -0,0 +1,541 @@
...
+    def inference(
+        self,
+        x: torch.Tensor,
+        y: torch.Tensor,
+        batch_index: Optional[torch.Tensor] = None,
+        label: Optional[torch.Tensor] = None,
+        n_samples=1,
+    ) -> Dict[str, Union[torch.Tensor, Dict[str, torch.Tensor]]]:
+        """ Internal helper function to compute necessary inference quanitities
```
[the remark this answers is not in the document]
Comment (adamgayoso) [reply scvi/models/totalvi.py]: done.

File: scvi/inference/total_inference.py [hunk 16]
```diff
@@ -0,0 +1,608 @@
...
+        ll_gene, ll_protein = self.compute_reconstruction_error(self.model)
+        return ll_gene, ll_protein
+
+    @torch.no_grad()
+    def marginal_ll(self, n_mc_samples=1000):
+        ll = self.compute_marginal_log_likelihood()
+        return ll
+
+    @torch.no_grad()
+    def get_background_mean(self):
```
[the remark this answers is not in the document]
Comment (adamgayoso) [reply scvi/inference/total_inference.py]: done.

File: scvi/dataset/dataset10X.py [hunk 17]
```diff
@@ -204,7 +204,8 @@ def populate(self):
                         measurement_type = "protein_expression"
                         columns_attr_name = "protein_names"
                         # protein counts are inherently not sparse
-                        measurement_data = measurement_data.A
+                        if self.dense is not True:
```
[the remark this answers is not in the document]
Comment (adamgayoso) [reply scvi/dataset/dataset10X.py]: we always make the protein counts dense as they do not contain many zeros, so  no advantage for sparse array. I updated the comment.

File: scvi/models/totalvi.py [hunk 18]
```diff
@@ -0,0 +1,541 @@
...
+        :param batch_index: array that indicates which batch the cells belong to with shape ``batch_size``
+        :param label: tensor of cell-types labels with shape ``(batch_size, n_labels)``
+        :param n_samples: number of samples
+        :return: tensor of predicted frequencies of expression with shape ``(batch_size, n_input)``
+        :rtype: :py:class:`torch.Tensor`
+        """
+        px_scale = self.inference(
+            x, y, batch_index=batch_index, label=label, n_samples=n_samples
+        )["px_scale"]
+        px_scale = torch.cat((px_scale["gene"], px_scale["protein"]), dim=-1)
```
[the remark this answers is not in the document]
Comment (adamgayoso) [reply scvi/models/totalvi.py]: this is so I can use the posterior differential expression score without having to rewrite it

File: scvi/models/modules.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (adamgayoso) [reply scvi/models/modules.py]: I'm going to close this for now, but we can rename it as we see fit in the future.

File: scvi/inference/total_inference.py [hunk 19]
```diff
@@ -0,0 +1,598 @@
...
+            )
+            rate = torch.cat((px_["rate"], protein_rate), dim=-1)
+            if len(px_["r"].size()) == 2:
+                px_dispersion_gene = px_["r"]
+            else:
+                px_dispersion_gene = torch.ones_like(x) * px_["r"]
+            if len(py_["r"].size()) == 2:
+                px_dispersion_protein = py_["r"]
+            else:
+                px_dispersion_protein = torch.ones_like(y) * py_["r"]
```
Comment (romain-lopez) [inline scvi/inference/total_inference.py]: px_dispersion_PROTEIN ?

File: scvi/inference/total_inference.py [hunk 20]
```diff
@@ -0,0 +1,598 @@
...
+
+            posterior_list[-1] = np.transpose(posterior_list[-1], (1, 2, 0))
+
+        return (
+            np.concatenate(posterior_list, axis=0),
+            np.concatenate(original_list, axis=0),
+        )
+
+    @torch.no_grad()
+    def get_sample_dropout(self, n_samples: int = 1, give_mean: bool = True):
```
Comment (romain-lopez) [inline scvi/inference/total_inference.py]: can we just docstring this is the zero inflation weight ? Just to be perfectly clear

File: scvi/models/totalvi.py [hunk 21]
```diff
@@ -0,0 +1,446 @@
...
+        if self.log_variational:
+            x = torch.log(1 + x)
+            y = torch.log(1 + y)
+        qz_m, qz_v, _, _, latent, _, = self.encoder(torch.cat((x, y), dim=-1))
+        z = latent["z"]
+        batch_index = fixed_batch * torch.ones_like(x[:, [0]])
+        # dummy library size as it's irrelevant here
+        library = 4.0 * torch.ones_like(x[:, [0]])
+        px_, py_ = self.decoder(z, library, batch_index)[0:2]
+        scale = torch.cat((px_["scale"], py_["scale"]), dim=-1)
```
Comment (romain-lopez) [inline scvi/models/totalvi.py]: This is still concatenated and not as a tuple
